connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaTopicBasedBackingStore.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.connect.storage;
+
+import org.apache.kafka.clients.admin.AdminClientConfig;
+import org.apache.kafka.clients.admin.NewTopic;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.common.config.TopicConfig;
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.connect.runtime.WorkerConfig;
+import org.apache.kafka.connect.util.Callback;
+import org.apache.kafka.connect.util.KafkaBasedLog;
+import org.apache.kafka.connect.util.TopicAdmin;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+public abstract class KafkaTopicBasedBackingStore {
```

Comment:
Is inheritance necessary here? It seems like we might make this a standalone class that can be composed into the various `Kafka*BackingStore` classes (and then possibly others) instead of only making it available through subclassing.

(Not a blocker)

Reply:
I tried it.
in one hand:
It'd require the composed field to be static, because of static methods would use the topicInitializer method:
https://github.com/apache/kafka/blob/15dfe065fb0ef34145ec204123c592e468842de8/connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaOffsetBackingStore.java#L129

in the other hand (which relates to the first one a bit): I'm not sure if I could inject this composition nicely, so that I can test it (or mock part of it) easily.
Currently the tests are prepared in a way that the `createKafkaBasedLog` is the method of the actual backingstore class, it would need a quite bit of a refactor on the tests too if I want to avoid this inheritance :)

wdyt?

Reply:
Ah, I forgot about the static constructors for the offset backing store. Yeah, probably not worth it to add another constructor variant for everything.

I was thinking about isolating this logic into a `Supplier<KafkaBasedLog>`, and we could then have something like:

```java
public class LogBuilder {
  public static Supplier<KafkaBasedLog<K, V>> createLog(String topic, String topicConfig, String topicPurpose...) {
    return () -> {
      KafkaBasedLog<K, V> result = ...
      // create and set up KafkaBasedLog, including performing topic initialization here
      return result;
    }
  }
}

public class KafkaConfigBackingStore implements ConfigBackingStore {
  public KafkaConfigBackingStore(Supplier<KafkaBasedLog> setupAndCreateKafkaBasedLog, ...)
}
```

Which I think would still technically work here (and honestly be a bit cleaner than how we're hacking the `setupAndCreateKafkaBasedLog` method in tests right now), but given the number of new constructors it would require (which introduce their own amount of ugliness), we can stick with what's in this PR right now.